### DISCUSSION TITLE
Question about the process of parsing the arguments input to `propagate()` and lifting tensors?

### DISCUSSION THREAD

**Discussion by alantes** [pyg-team/pytorch_geometric#6447] [Q&A] [answered]
Hi folks, thank you so much for developing PyG.
When I read the source code in `propagate()` and `__collect__()` in `message_passing.py` and  set breakpoints along the way, I found `self.__user_args__` really making me confused.
For example, if I input an tensor `x` (shape: `[2190, 2]`) into `propagate()` as a kwarg, when executing `coll_dict = self.__collect__(self.__user_args__, edge_idex, size, kwargs)`, the `self.__user_args__` already contains `x_j`. But I do not find a place in `MessagePassing` that you change `self.__user_args__`, except for `__init__()`. But when executing `__init__()`, `x` is not even introduced.
I wonder where in the program do you parse the arguments input to `propagate()`, and how do you judge whether or not an argument can be lifted to `_i` and `_j` components? Through dimensions?
Thank you in advance. I will appreciate your help a lot.

**Comment by rusty1s** [comment, marked as answer]
`__user_args__` collects the arguments of `message`, `aggregate` and `update` and collects them. By default, `message` is implemented as
```python
def message(self, x_j: Tensor):
     return x_j
```
That's why it will contain `x_j` without any custom `message` definition.

**Reply by alantes** [reply]
That's very helpful. Thank you for your reply.
